### DISCUSSION TITLE
Positional DML statement longevity

### DISCUSSION THREAD

**Discussion by aafemt** [FirebirdSQL/firebird#8196] [General]
If a positional DML statement is prepared and then referenced cursor is closed and reopened, what should be the behavior of a following attempt to execute the prepared statement?

Should it throw an error?
Should it be executed ok if the new cursor is compatible with the old one (i.e. includes the same table)?

**Comment by asfernandes** [comment]
I believe you should always start saying what is the current behavior and if it was changed anytime, as the current/old behavior will always have importance on the decision if something already existing is wrong or correct.

**Reply by aafemt** [reply]
Current behavior is throwing error. I failed to locate positioned DML code in FB 2.1 sources so I cannot tell how it worked back then.

Because it is rather a question of design and UX, I would prefer to disregard current and past behaviors concentrating on the usability of the feature.

**Comment by dyemanov** [comment]
In the ideal world, I suppose it could be executed again without errors. The question is whether it's appropriate to move the cursor compatibility check from the prepare time to the execution time.

**Reply by hvlad** [reply]
Could it be in both places ?

**Reply by aafemt** [reply]
Sure it could. I've almost done it [there](https://github.com/aafemt/firebird/blob/6c39e6b439c8a8e79b08b84f3ef1a7fd700583ef/src/dsql/DsqlRequests.cpp). Routine `mapCursorKey()` instead of throwing error immediately can easily relink parent cursor if possible.
